File: go/ring.go
package consistenthash

import (
	"fmt"
	"sort"
)

// Ring is consistent hashing with virtual nodes: keys and nodes share one circular
// FNV space, and a key is owned by the first node clockwise from it. Each physical
// node occupies Replicas points on the ring so load spreads evenly and a departing
// node's keys scatter across many neighbors.
type Ring struct {
	replicas int
	ring     map[uint32]string
	sorted   []uint32
	nodes    map[string]struct{}
}

// NewRing builds a ring with the given nodes and virtual-node count (replicas).
func NewRing(nodes []string, replicas int) *Ring {
	if replicas < 1 {
		panic("replicas must be at least 1")
	}
	r := &Ring{
		replicas: replicas,
		ring:     make(map[uint32]string),
		nodes:    make(map[string]struct{}),
	}
	for _, n := range nodes {
		r.Add(n)
	}
	return r
}

// Add inserts a node and its virtual replicas. A repeated node is a no-op.
func (r *Ring) Add(node string) {
	if _, ok := r.nodes[node]; ok {
		return
	}
	r.nodes[node] = struct{}{}
	for i := 0; i < r.replicas; i++ {
		slot := Fnv1a32(fmt.Sprintf("%s#%d", node, i))
		for {
			if _, taken := r.ring[slot]; !taken {
				break
			}
			slot++ // deterministic nudge on collision
		}
		r.ring[slot] = node
	}
	r.reindex()
}

// Remove deletes a node and all its slots. Removing an absent node is a no-op.
func (r *Ring) Remove(node string) {
	if _, ok := r.nodes[node]; !ok {
		return
	}
	delete(r.nodes, node)
	for slot, n := range r.ring {
		if n == node {
			delete(r.ring, slot)
		}
	}
	r.reindex()
}

// Get returns the node that owns key, or "" if the ring is empty.
func (r *Ring) Get(key string) string {
	if len(r.sorted) == 0 {
		return ""
	}
	return r.ring[r.sorted[r.firstClockwise(Fnv1a32(key))]]
}

// GetReplicas returns the first count distinct nodes clockwise from key.
func (r *Ring) GetReplicas(key string, count int) []string {
	result := []string{}
	if len(r.sorted) == 0 || count <= 0 {
		return result
	}
	start := r.firstClockwise(Fnv1a32(key))
	cap := count
	if cap > len(r.nodes) {
		cap = len(r.nodes)
	}
	seen := make(map[string]struct{})
	for i := 0; i < len(r.sorted); i++ {
		node := r.ring[r.sorted[(start+i)%len(r.sorted)]]
		if _, dup := seen[node]; dup {
			continue
		}
		seen[node] = struct{}{}
		result = append(result, node)
		if len(result) == cap {
			break
		}
	}
	return result
}

// Nodes returns the physical nodes in sorted order.
func (r *Ring) Nodes() []string {
	return sortedKeys(r.nodes)
}

// Len is the number of physical nodes.
func (r *Ring) Len() int { return len(r.nodes) }

func (r *Ring) firstClockwise(h uint32) int {
	idx := sort.Search(len(r.sorted), func(i int) bool { return r.sorted[i] > h })
	if idx == len(r.sorted) {
		return 0 // wrap around
	}
	return idx
}

func (r *Ring) reindex() {
	r.sorted = r.sorted[:0]
	for slot := range r.ring {
		r.sorted = append(r.sorted, slot)
	}
	sort.Slice(r.sorted, func(i, j int) bool { return r.sorted[i] < r.sorted[j] })
}

func sortedKeys(m map[string]struct{}) []string {
	out := make([]string, 0, len(m))
	for k := range m {
		out = append(out, k)
	}
	sort.Strings(out)
	return out
}

File: rust/src/ring.rs
use crate::fnv1a_32;
use std::collections::{BTreeSet, HashMap};

/// A consistent-hash ring with virtual nodes. Keys and nodes share one circular FNV
/// space; a key is owned by the first node clockwise from it. Each node occupies
/// `replicas` ring points so load spreads evenly.
pub struct Ring {
    replicas: usize,
    ring: HashMap<u32, String>,
    sorted: Vec<u32>,
    nodes: BTreeSet<String>,
}

impl Ring {
    pub fn new(nodes: &[&str], replicas: usize) -> Self {
        assert!(replicas >= 1, "replicas must be at least 1");
        let mut r = Ring {
            replicas,
            ring: HashMap::new(),
            sorted: Vec::new(),
            nodes: BTreeSet::new(),
        };
        for n in nodes {
            r.add(n);
        }
        r
    }

    pub fn add(&mut self, node: &str) {
        if !self.nodes.insert(node.to_string()) {
            return;
        }
        for i in 0..self.replicas {
            let mut slot = fnv1a_32(&format!("{node}#{i}"));
            while self.ring.contains_key(&slot) {
                slot = slot.wrapping_add(1); // deterministic nudge on collision
            }
            self.ring.insert(slot, node.to_string());
        }
        self.reindex();
    }

    pub fn remove(&mut self, node: &str) {
        if !self.nodes.remove(node) {
            return;
        }
        self.ring.retain(|_, n| n != node);
        self.reindex();
    }

    pub fn get(&self, key: &str) -> Option<&str> {
        if self.sorted.is_empty() {
            return None;
        }
        let idx = self.first_clockwise(fnv1a_32(key));
        Some(self.ring[&self.sorted[idx]].as_str())
    }

    pub fn get_replicas(&self, key: &str, count: usize) -> Vec<String> {
        let mut result = Vec::new();
        if self.sorted.is_empty() || count == 0 {
            return result;
        }
        let start = self.first_clockwise(fnv1a_32(key));
        let cap = count.min(self.nodes.len());
        for i in 0..self.sorted.len() {
            let node = &self.ring[&self.sorted[(start + i) % self.sorted.len()]];
            if !result.iter().any(|n| n == node) {
                result.push(node.clone());
                if result.len() == cap {
                    break;
                }
            }
        }
        result
    }

    pub fn nodes(&self) -> Vec<String> {
        self.nodes.iter().cloned().collect()
    }

    pub fn len(&self) -> usize {
        self.nodes.len()
    }

    pub fn is_empty(&self) -> bool {
        self.nodes.is_empty()
    }

    fn first_clockwise(&self, h: u32) -> usize {
        match self.sorted.partition_point(|&s| s <= h) {
            i if i == self.sorted.len() => 0, // wrap around
            i => i,
        }
    }

    fn reindex(&mut self) {
        self.sorted = self.ring.keys().copied().collect();
        self.sorted.sort_unstable();
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn adding_a_node_remaps_a_small_fraction() {
        let keys: Vec<String> = (0..5000).map(|i| format!("key-{i}")).collect();
        let mut r = Ring::new(&["a", "b", "c"], 200);
        let before: Vec<String> = keys.iter().map(|k| r.get(k).unwrap().to_string()).collect();
        r.add("d");
        let mut moved = 0;
        for (i, k) in keys.iter().enumerate() {
            let after = r.get(k).unwrap();
            if after != before[i] {
                moved += 1;
                assert_eq!(after, "d", "moved key must land on the new node");
            }
        }
        assert!((moved as f64) / (keys.len() as f64) < 0.40);
    }

    #[test]
    fn replicas_are_distinct() {
        let r = Ring::new(&["a", "b", "c", "d", "e"], 100);
        let reps = r.get_replicas("some-key", 3);
        assert_eq!(reps.len(), 3);
        let unique: BTreeSet<_> = reps.iter().collect();
        assert_eq!(unique.len(), 3);
    }
}
